### ISSUE TITLE
Feature: MegaLoc forge (viper.forge package)

### ISSUE THREAD

**Issue by markvilar** [markvilar/viper#11]
## Problem and motivation

Issue #6 added the generic `viper.model_decomposition` primitives (SVD-based linear truncation) but deliberately left model-specific application out of scope. Those primitives operate on raw tensors and `nn.Linear` modules with no knowledge of any architecture, so on their own they cannot produce a usable, lower-dimensional MegaLoc: something has to know *which* layer of MegaLoc to decompose (the aggregator's final projection) and how to recompose the module tree afterwards.

MegaLoc emits 8448D descriptors from a final linear projection of roughly `(8448, 16640)`, which both dominates the model's weight memory and is poorly conditioned for the hyperspherical probabilistic modelling viper targets (see #6 for the full argument). We need a MegaLoc-specific "forge" that applies the #6 primitives to the right layer and rebuilds the model — geometrically close to the original at initialisation, so downstream fine-tuning does not cold-start.

This issue introduces the `viper.forge` sub-package (the model-specific "forge" layer anticipated in #6) via its first implementation — MegaLoc. Producing and publishing the actual truncated checkpoints is deferred to a later issue, once the model-adapter CLI (#17) exists to bake them ergonomically rather than via a throwaway script.

## Proposed solution

Add a `viper.forge` sub-package: model-specific builders that apply the generic `model_decomposition` primitives to a given architecture and return a rebuilt model. It is organised one module per model, mirroring `viper.models`; the model carries the structural knowledge, while the decomposition *method* comes from the #6 primitives.

```
src/viper/forge/
    __init__.py     # public re-exports
    megaloc.py      # MegaLoc forge function(s)
```

The MegaLoc forge is a thin recomposition over the existing `truncate_linear` primitive — the final projection lives at `model.aggregator.linear` (a plain `nn.Linear`), and `MegaLocAggregationModule` / `MegaLocModel` are simple composition constructors. It takes a loaded model and returns a new one:

```python
# viper/forge/megaloc.py
def forge_megaloc_svd_truncated(model: MegaLocModel, k: int) -> MegaLocModel:
    """
    Rebuild MegaLoc with its aggregator's final projection SVD-truncated to k
    output dimensions.

    Returns a new MegaLocModel; the input is not mutated. The backbone and SALAD
    aggregator are reused unchanged, and only the final linear projection is
    replaced with its truncated-SVD re-expression (see truncate_linear). Device
    and dtype are inherited from the source model. Descriptor dimension of the
    result is k.
    """
    truncated_linear = truncate_linear(model.aggregator.linear, k)
    aggregator = MegaLocAggregationModule(
        salad=model.aggregator.salad, linear=truncated_linear
    )
    return MegaLocModel(backbone=model.backbone, aggregator=aggregator)
```

Design decisions:

- **Model in, model out.** The `model_decomposition` primitives are module-level and `MegaLocModel` is already the composition unit, so the forge is pure recomposition — no re-deriving from raw tensors. This also drops straight into the CLI (#17), which resolves models to instances via the registry.
- **Serialization stays separate.** The forge does exactly one thing; producing a `state_dict()` and writing a checkpoint belongs to the publishing step (deferred issue), not here.
- **Non-destructive.** `truncate_linear` returns a fresh `nn.Linear`; the backbone and SALAD are reused by reference, leaving the source model intact.
- **No re-validation of `k`.** `truncate_linear` → `truncate_linear_svd` already enforces `0 < k <= out_features` with a clear error; the forge delegates.
- **Device-agnostic.** Truncation inherits the source layer's device/dtype, so the forge runs equally on CPU or CUDA.

Validated by unit tests rather than by baking a checkpoint: the forged model emits `k`-dimensional descriptors, `vector_size == k`, the source model is left unmutated, and (for `k == vector_size`) the L2-normalised output stays geometrically close to the original.

## Open questions

None outstanding — the input/output shape is settled (model in, model out) and checkpoint production is deferred to the publishing issue.

## Notes

- Introduces the `viper.forge` package via its first implementation; further models get their own modules in later issues.
- The forge *registry* / dispatch layer is out of scope here — its only consumer is the CLI (#17), which introduces it. This issue calls the forge function directly.
- Producing and publishing the SVD-truncated MegaLoc checkpoints is deferred to a follow-up issue that depends on the CLI (#17); the file-naming convention and the choice of output dimension(s) `k` live there.
- Related: #6 (decomposition primitives, done — direct dependency), #17 (model-adapter CLI that drives forges; megaloc is its first entry), #15 (expose truncated variants by registry key).
